src/box/allocator.h

```diff
@@ -0,0 +1,229 @@
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <tuple>
+
+#include <small/small.h>
+#include "sysalloc.h"
+
+struct alloc_stat {
+	size_t used;
+	size_t total;
+};
+
+struct allocator_stats {
+	struct alloc_stat small;
+	struct alloc_stat sys;
+};
+
+int
+small_stats_noop_cb(const struct mempool_stats *stats, void *cb_ctx);
+
+struct allocator_settings {
+	struct small_allocator {
+		struct slab_cache *cache;
+		uint32_t objsize_min;
+		unsigned granularity;
+		float alloc_factor;
+		float *actual_alloc_factor;
+	} small;
+	struct system_allocator {
+		struct quota *quota;
+	} sys;
+};
+
+static inline void
+allocator_settings_init(allocator_settings *settings, struct slab_cache *cache,
+			uint32_t objsize_min, unsigned granularity,
+			float alloc_factor, float *actual_alloc_factor,
+			struct quota *quota)
+{
+	settings->small.cache = cache;
+	settings->small.objsize_min = objsize_min;
+	settings->small.granularity = granularity;
+	settings->small.alloc_factor = alloc_factor;
+	settings->small.actual_alloc_factor = actual_alloc_factor;
+
+	settings->sys.quota = quota;
+}
+
+/**
+ * Each allocator class should feature at least following interfaces:
+ *
+ * void create(struct allocator_settings *settings);
+ * void destroy(void);
+ * void* alloc(size_t size);
+ * void free(void* ptr, size_t size);
+ * void stats(struct alloc_stats);
+ */
+class SmallAlloc
+{
+public:
+	static inline void
+	create(struct allocator_settings *settings)
+	{
+		small_alloc_create(&small_alloc,
+				   settings->small.cache,
+				   settings->small.objsize_min,
+				   settings->small.granularity,
+				   settings->small.alloc_factor,
+				   settings->small.actual_alloc_factor);
+	}
+	static inline void
+	destroy(void)
+	{
+		small_alloc_destroy(&small_alloc);
+	}
+	static inline void *
+	alloc(size_t size)
+	{
+		return smalloc(&small_alloc, size);
+	}
+	static inline void
+	free(void *ptr, size_t size)
+	{
+		return smfree(&small_alloc, ptr, size);
+	}
+	static inline void
+	stats(struct allocator_stats *alloc_stats, mempool_stats_cb cb,
+	      void *cb_ctx)
+	{
+		struct small_stats data_stats;
+		small_stats(&SmallAlloc::small_alloc, &data_stats,
+			    cb, cb_ctx);
+		alloc_stats->small.used = data_stats.used;
+		alloc_stats->small.total = data_stats.total;
+	}
+	static inline struct small_alloc *
+	get_alloc(void)
+	{
+		return &small_alloc;
+	}
+private:
+	static struct small_alloc small_alloc;
+};
+
+class SysAlloc
+{
+public:
+	static inline void
+	create(struct allocator_settings *settings)
+	{
+		sys_alloc_create(&sys_alloc, settings->sys.quota);
+	}
+	static inline void
+	destroy(void)
+	{
+		sys_alloc_destroy(&sys_alloc);
+	}
+	static inline void *
+	alloc(size_t size)
+	{
+		return sysalloc(&sys_alloc, size);
+	}
+	static inline void
+	free(void *ptr, size_t size)
+	{
+		return sysfree(&sys_alloc, ptr, size);
+	}
+	static inline void
+	stats(struct allocator_stats *alloc_stats, mempool_stats_cb cb,
```

Comment:
I think that it is not good to pass `mempool_stats_cb` function as argument for `stats` function of SysAlloc. Mempool stats not relevant for it. Also later if we implement other allocator which will require some callback function for `stats` function, we must be able to pass a callback function with any signature.